src/stream/src/common/log_store/kv_log_store/serde.rs

```diff
@@ -0,0 +1,598 @@
+        let value_bytes = self.row_serde.serialize(extended_row).into();
+        (key_bytes, value_bytes)
+    }
+
+    pub fn serialize_epoch(&self, vnode: VirtualNode, epoch: u64) -> Bytes {
+        serialize_pk_with_vnode(
+            [Some(ScalarImpl::Int64(Self::encode_epoch(epoch)))],
+            &self.epoch_serde,
+            vnode,
+        )
+    }
+
+    pub fn serialize_log_store_pk(
+        &self,
+        vnode: VirtualNode,
+        epoch: u64,
+        seq_id: SeqIdType,
+    ) -> Bytes {
+        serialize_pk_with_vnode(
+            [
+                Some(ScalarImpl::Int64(Self::encode_epoch(epoch))),
+                Some(ScalarImpl::Int32(seq_id)),
+            ],
+            &self.pk_serde,
+            vnode,
+        )
+    }
+
+    pub fn serialize_truncation_offset_watermark(
+        &self,
+        vnode: VirtualNode,
+        offset: ReaderTruncationOffsetType,
+    ) -> Bytes {
+        let curr_offset = self.serialize_epoch(vnode, offset);
+        let ret = Bytes::from(next_key(&curr_offset));
+        assert!(!ret.is_empty());
+        ret
+    }
+}
+
+impl LogStoreRowSerde {
+    fn deserialize(&self, value_bytes: Bytes) -> LogStoreResult<(u64, LogStoreRowOp)> {
+        let row_data = self.row_serde.deserialize(&value_bytes)?;
+
+        let payload_row = OwnedRow::new(row_data[PREDEFINED_COLUMNS_TYPES.len()..].to_vec());
+        let epoch = Self::decode_epoch(*row_data[EPOCH_INDEX].as_ref().unwrap().as_int64());
+        let row_op_code = *row_data[ROW_OP_INDEX].as_ref().unwrap().as_int16();
+
+        let op = match row_op_code {
+            INSERT_OP_CODE => LogStoreRowOp::Row {
+                op: Op::Insert,
+                row: payload_row,
+            },
+            DELETE_OP_CODE => LogStoreRowOp::Row {
+                op: Op::Delete,
+                row: payload_row,
+            },
+            UPDATE_INSERT_OP_CODE => LogStoreRowOp::Row {
+                op: Op::UpdateInsert,
+                row: payload_row,
+            },
+            UPDATE_DELETE_OP_CODE => LogStoreRowOp::Row {
+                op: Op::UpdateDelete,
+                row: payload_row,
+            },
+            BARRIER_OP_CODE => {
+                assert!(row_data[SEQ_ID_INDEX].is_none());
+                LogStoreRowOp::Barrier {
+                    is_checkpoint: false,
+                }
+            }
+            CHECKPOINT_BARRIER_OP_CODE => {
+                assert!(row_data[SEQ_ID_INDEX].is_none());
+                LogStoreRowOp::Barrier {
+                    is_checkpoint: true,
+                }
+            }
+            _ => unreachable!("invalid row op code: {}", row_op_code),
+        };
+        Ok((epoch, op))
+    }
+
+    pub async fn deserialize_stream_chunk(
+        &self,
+        stream: impl StateStoreReadIterStream,
+        start_seq_id: SeqIdType,
+        end_seq_id: SeqIdType,
+        expected: u64,
+    ) -> LogStoreResult<StreamChunk> {
```

Comment:
Does it mean that the upper layer needs to maintain `seq_id` to deserialize a stream? How do we get the `start_seq_id` and `end_seq_id` on recovery?

Reply:
+1. I have the same question. I think LogReader should maintain the per epoch seq_id and persist it in a state table?

Reply:
The `start_seq_id` and `end_seq_id` will be maintained in memory, and is used to rebuild a stream chunk from the state store. During recovery, we do range scan with the left bound as `Unbounded`, which means during recovery, we read data from the oldest untruncated record. This is still correct when we only ensure at-least-once delivery.